app/scripts/modules/core/instance/instances.component.ts

```diff
@@ -106,6 +106,10 @@ export class InstancesController implements IComponentController {
     this.$scope.$watch('highlight', () => this.renderInstances());
   }
 
+  public $onChanges(): void {
```

Comment:
careful with this - it will actually get called twice when the component renders.

Should probably be rewritten as:
```typescript

interface IOnChanges extends IOnChangesObject {
  instances: IChangesObject<Instance[]>;
  highlight: IChangesObject<string>;
}
...
  public $onChanges(changes: IOnChanges): void {
    if (changes.instances && !changes.instances.isFirstChange()) {
      this.renderInstances();
    }
  }
```
This will avoid an initial extra string build in the render method and shave some time off the paint (try it on a very large edge service to get a feel for the performance difference).

You could make a few other changes to the `renderInstances` method that may or may not make some difference in performance:
1. instead of `sortBy(this.instances, 'launchTime')`, skip the lodash: 
  `const instances = (this.instances || []).sort((a, b) => a.launchTime - b.launchTime);`
2. instead of constantly calling `$state.includes` to see if the instance is active, grab the `provider` and `instanceId` from `$stateParams`...okay, fine, here is what I did locally:
```typescript
private renderInstances(): void {
    this.activeInstance = null;
    const instances = (this.instances || []).sort((a, b) => a.launchTime - b.launchTime);
    const showingDetails = this.$state.current.name.endsWith('.instanceDetails'),
          showingId = this.$stateParams.instanceId,
          showingProvider = this.$stateParams.provider;

    const innerHtml = '<div class="instances">' + instances.map((instance) => {
        const id = instance.id;
        let activeClass = '';

        if (showingDetails && showingId === instance.id && showingProvider === instance.provider) {
          activeClass = ' active';
          this.activeInstance = { instanceId: instance.id, provider: instance.provider };
        }
        if (this.highlight === id) {
          activeClass += ' highlighted';
        }

        return '<a title="' + id +
          '" data-provider="' + instance.provider +
          '" data-toggle="tooltip" data-instance-id="' + id +
          '" class="instance health-status-' + instance.healthState + activeClass + '"></a>';
      }).join('') + '</div>';

    if (innerHtml !== this.$element.get(0).innerHTML) {
      this.removeTooltips();
      this.$element.get(0).innerHTML = innerHtml;
    }
  }
```
That should cut the render for very large services by about half.